File: src/meshplanner/sites/grid.py
"""Generate regular grid of candidate sites.

Converts km spacing to lat/lon step sizes using the haversine approximation:
    - 1 degree latitude  ≈ 111.32 km (constant)
    - 1 degree longitude ≈ 111.32 * cos(lat) km (depends on latitude)
"""

import math

from shapely.geometry import Point, Polygon

from meshplanner.sites.candidate import CandidateSite


def generate_grid(
    bbox: dict[str, float],
    spacing_km: float = 1.0,
    name_prefix: str = "Grid",
) -> list[CandidateSite]:
    """Generate a regular grid of candidate sites within a bounding box.

    Args:
        bbox: Bounding box with keys {"west", "south", "east", "north"} (decimal degrees).
        spacing_km: Target spacing between adjacent sites in kilometres.
        name_prefix: Prefix for site names (default "Grid").

    Returns:
        List of CandidateSite objects named "{prefix}-{row}-{col}".
    """
    west = bbox["west"]
    south = bbox["south"]
    east = bbox["east"]
    north = bbox["north"]

    # Use the midpoint latitude for longitude step calculation
    mid_lat = (south + north) / 2.0
    lat_rad = math.radians(mid_lat)

    lat_step = spacing_km / 111.32
    lon_step = spacing_km / (111.32 * math.cos(lat_rad))

    # Guard against degenerate bounding boxes
    if lat_step <= 0 or lon_step <= 0:
        return []
    if north <= south or east <= west:
        return []

    sites: list[CandidateSite] = []

    # Generate rows from south to north
    lat = south
    row = 0
    while lat <= north:
        lon = west
        col = 0
        while lon <= east:
            name = f"{name_prefix}-{row}-{col}"
            sites.append(CandidateSite(lat=lat, lon=lon, name=name))
            lon += lon_step
            col += 1
        lat += lat_step
        row += 1

    return sites


def generate_grid_within_polygon(
    polygon_coords: list[tuple[float, float]],
    spacing_km: float = 1.0,
) -> list[CandidateSite]:
    """Generate a regular grid of candidate sites within an arbitrary polygon.

    A bounding-box grid is created first, then sites falling outside the
    polygon are filtered out via shapely's ``Point.within()`` test.

    Args:
        polygon_coords: List of (longitude, latitude) tuples defining the polygon
                        boundary (in clockwise or counter-clockwise order).
        spacing_km: Target spacing between adjacent sites in kilometres.

    Returns:
        List of CandidateSite objects that lie inside the polygon.
    """
    if len(polygon_coords) < 3:
        return []

    polygon = Polygon(polygon_coords)

    # Compute bounding box of the polygon
    lons = [p[0] for p in polygon_coords]
    lats = [p[1] for p in polygon_coords]
    bbox = {
        "west": min(lons),
        "south": min(lats),
        "east": max(lons),
        "north": max(lats),
    }

    # Generate full grid within the bbox
    all_sites = generate_grid(bbox, spacing_km=spacing_km, name_prefix="Grid")

    # Keep only sites inside the polygon
    sites: list[CandidateSite] = []
    for site in all_sites:
        point = Point(site.lon, site.lat)
        if point.within(polygon):
            sites.append(site)

    return sites

File: meshplanner-app/src/lib/sites/grid.ts
/**
 * Generate regular grids of candidate sites within a bounding box or polygon.
 *
 * Converts km spacing to lat/lon step sizes using the haversine approximation:
 *   - 1 degree latitude  ≈ 111.32 km (constant)
 *   - 1 degree longitude ≈ 111.32 * cos(lat) km (depends on latitude)
 *
 * Ported from Python's src/meshplanner/sites/grid.py.
 */

import type { CandidateSite, Bbox } from '../types'

const KM_PER_DEG_LAT = 111.32

function degToRad(deg: number): number {
  return (deg * Math.PI) / 180
}

/**
 * Convert a km spacing to lat/lon degree steps at a given reference latitude.
 */
function kmToDegSteps(
  spacingKm: number,
  latitude: number,
): { latStep: number; lonStep: number } {
  const latStep = spacingKm / KM_PER_DEG_LAT
  const lonStep = spacingKm / (KM_PER_DEG_LAT * Math.cos(degToRad(latitude)))
  return { latStep, lonStep }
}

/**
 * Ray-casting point-in-polygon test.
 *
 * Returns `true` when the point (lon, lat) lies inside the polygon defined by
 * `polygon` — an array of `[longitude, latitude]` pairs (matching the
 * convention used by the Python shapely code).
 */
function pointInPolygon(
  lon: number,
  lat: number,
  polygon: [number, number][],
): boolean {
  let inside = false
  const n = polygon.length
  for (let i = 0, j = n - 1; i < n; j = i++) {
    const [xi, yi] = polygon[i]!
    const [xj, yj] = polygon[j]!
    if (yi > lat !== yj > lat &&
        lon < ((xj - xi) * (lat - yi)) / (yj - yi) + xi) {
      inside = !inside
    }
  }
  return inside
}

/**
 * Generate a regular grid of candidate sites within a bounding box.
 *
 * @param bbox        Bounding box with keys `west`, `south`, `east`, `north`.
 * @param spacingKm   Target spacing between adjacent sites in kilometres
 *                    (default 1.0).
 * @param namePrefix  Prefix for site names (default "Grid").
 * @returns           List of CandidateSite objects named
 *                    `"{prefix}-{row}-{col}"`.
 */
export function generateGrid(
  bbox: Bbox,
  spacingKm: number = 1.0,
  namePrefix: string = 'Grid',
): CandidateSite[] {
  const { west, south, east, north } = bbox

  // Guard against degenerate bounding boxes
  if (spacingKm <= 0 || north <= south || east <= west) {
    return []
  }

  // Use the midpoint latitude for longitude step calculation
  const midLat = (south + north) / 2.0
  const { latStep, lonStep } = kmToDegSteps(spacingKm, midLat)

  if (latStep <= 0 || lonStep <= 0) {
    return []
  }

  const sites: CandidateSite[] = []

  // Generate rows from south to north, columns from west to east
  let lat = south
  let row = 0
  while (lat <= north + 1e-12) {
    let lon = west
    let col = 0
    while (lon <= east + 1e-12) {
      sites.push({
        name: `${namePrefix}-${row}-${col}`,
        latitude: lat,
        longitude: lon,
      })
      lon += lonStep
      col++
    }
    lat += latStep
    row++
  }

  return sites
}

/**
 * Generate a regular grid of candidate sites within an arbitrary polygon.
 *
 * A bounding-box grid is created first, then sites that fall outside the
 * polygon are filtered out via a ray-casting point-in-polygon test.
 *
 * @param polygonCoords  Array of `[longitude, latitude]` pairs defining the
 *                       polygon boundary (clockwise or counter-clockwise).
 * @param spacingKm      Target spacing between adjacent sites in kilometres
 *                       (default 1.0).
 * @returns              List of CandidateSite objects that lie inside the
 *                       polygon.
 */
export function generateGridWithinPolygon(
  polygonCoords: [number, number][],
  spacingKm: number = 1.0,
): CandidateSite[] {
  if (polygonCoords.length < 3) {
    return []
  }

  // Compute bounding box of the polygon
  let west = Infinity
  let south = Infinity
  let east = -Infinity
  let north = -Infinity

  for (const [lon, lat] of polygonCoords) {
    if (lon < west) west = lon
    if (lat < south) south = lat
    if (lon > east) east = lon
    if (lat > north) north = lat
  }

  const bbox: Bbox = { west, south, east, north }

  // Generate full grid within the bbox, then filter
  const allSites = generateGrid(bbox, spacingKm, 'Grid')
  return allSites.filter(site =>
    pointInPolygon(site.longitude, site.latitude, polygonCoords),
  )
}
